zmq_actor.hpp

```diff
@@ -0,0 +1,99 @@
+/*
+    Copyright (c) 2020 ZeroMQ community
+    Copyright (c) 2020 Brett Viren
+
+    Permission is hereby granted, free of charge, to any person obtaining a copy
+    of this software and associated documentation files (the "Software"), to
+    deal in the Software without restriction, including without limitation the
+    rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
+    sell copies of the Software, and to permit persons to whom the Software is
+    furnished to do so, subject to the following conditions:
+
+    The above copyright notice and this permission notice shall be included in
+    all copies or substantial portions of the Software.
+
+    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+    IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+    FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+    AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+    LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+    FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+    IN THE SOFTWARE.
+*/
+
+#ifndef __ZMQ_ACTOR_HPP_INCLUDED__
+#define __ZMQ_ACTOR_HPP_INCLUDED__
+
+#include "zmq.hpp"
+
+#ifdef ZMQ_CPP11
+
+#include <thread>
+
+namespace zmq
+{
+/*! Return two paired PAIR sockets.
+   
+  First is bound and second is connected via inproc://.
+*/
+std::pair<zmq::socket_t, zmq::socket_t> create_linked_pairs(context_t &ctx)
+{
+    std::pair<zmq::socket_t, zmq::socket_t> ret{socket_t(ctx, socket_type::pair),
+                                                socket_t(ctx, socket_type::pair)};
+
+    std::stringstream ss;
+    ss << "inproc://paired-" << std::hex << ret.first.handle() << "-"
+       << ret.second.handle();
+    std::string addr = ss.str();
+    ret.first.bind(addr.c_str());
+    ret.second.connect(addr.c_str());
+    return ret;
+}
```

Comment:
May I suggest a more generic function?
```C++
void link_inproc(socket_ref first, socket_ref second)
{
    std::stringstream ss;
    ss << "inproc://paired-" << std::hex << first.handle() << "-"
       << second.handle();
    std::string addr = ss.str();
    first.bind(addr.c_str());
    second.connect(addr.c_str());
}
```

Reply:
That gives the user the option of setting socket options before binding/connecting.

Reply:
This is fine.  This function is needed (for now) only inside the `actor_t`.  I guess it could just as well be removed with the body residing in the `actor_t`.  Separating it to another function is influence from CZMQ.